vizro-core/src/vizro/models/_action/_action.py

```diff
@@ -59,6 +58,57 @@ def build(self):
             "action_finished": Output("action_finished", "data", allow_duplicate=True),
         }
 
+        action_components = _get_action_callback_mapping(
+            action_id=self.id, argument="components"  # type: ignore[arg-type]
+        )
+
+        return callback_inputs, callback_outputs, action_components
+
+    def _action_callback_function(self, **inputs: Dict[str, Any]) -> Dict[str, Any]:
+        logger.debug("=============== ACTION ===============")
+        logger.debug(f'Action ID: "{self.id}"')
+        logger.debug(f'Action name: "{self.function._function.__name__}"')
+        logger.debug(f"Action inputs: {inputs}")
+
+        # Invoking the action's function
+        return_value = self.function(**inputs) or {}
+
+        # Raising the custom exception if return value length doesn't match the number of outputs
+        return_value_len = (
+            1 if not hasattr(return_value, "__len__") or isinstance(return_value, str) else len(return_value)
+        )
+        outputs = list(ctx.outputs_grouping.keys())
+        outputs.remove("action_finished")
+        if len(outputs) != return_value_len:
+            raise ValueError(
+                f"Number of action's returned elements: {return_value_len} does not match the number"
+                f" of action's defined outputs: {len(outputs)}."
+            )
+
+        if isinstance(return_value, dict):
+            return {"action_finished": None, **return_value}
+
+        # If return_value is a single element, ensure return_value is a list
+        if not isinstance(return_value, (list, tuple)):
+            return_value = [return_value]
+
+        # Map returned values to dictionary format where None belongs to the "action_finished" output
+        return {"action_finished": None, **dict(zip(outputs, return_value))}
```

Comment:
For upcoming custom actions we want to support few different ways of defining return values (as Dash already does):
```
    @capture("action")
    def custom_action(input: int):x
        return input+1, input+2   # return value as a tuple 
        return [input+1, input+2]   # return value as a list
        return {"output_2_children": input+2, "output_1_children": input_1}.   # return value as a dict
``` 

So we want to support return value to be:
- single element,
- tuple of elements,
- list of elements,
- dictionary.

Regarding that we add "action_finished" internal `dash.Output`, we always pack the users custom action function return value to the dictionary format so we don't mess the order of outputs.

Reply:
Do we actually want to allow all these possibilities even if they're possible in Dash? Seems like dictionary is the most general one and doesn't leave any ambiguity in ordering arguments (plus we should allow single element so you don't need to write a dictionary just for one thing).

I remember thinking about this before and not sure whether we discussed it or not but interested in your thoughts on when the different output formats are useful. It's fine by me if we do want to allow the others, just would be good to understand. Tuple of elements makes sense to me because it's how you write normal Python functions; list of elements feels like kind of a weird Dash thing.

Reply:
Valid point. I concur that returning values like `dictionary`, `single element`, and `tuple` is clear. However, I'm now hesitant about using a `list` as a return value due to potential ambiguity. The question arises: Should we interpret the returned list as a single element, or is it meant to be treated differently? 

Since this logic is closely related to `custom actions` that are **not officially released**, do you mind if we continue this conversation in that PR (will open in October)? I'm also happy to re-implement the `list' handling in this PR if we have a strong preference for how to handle them.

Reply:
Very happy to continue the conversation elsewhere - it's not urgent at all and not related to this PR, I just commented on it since I happened to see it here 🙂